From 170dd08df66240c1755f8904b56dc4a825eb7fef Mon Sep 17 00:00:00 2001
From: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Date: Mon, 8 Jan 2024 15:57:30 +0100
Subject: [PATCH 077/120] remoteproc: add support of OP-TEE loading based on
 compatibility

Use compatibles defined in Linux and Op-TEE to specify the platform and
the loading method.
st,stm32mp1-m4: load of an ELF image on stm32mp1 Cortex-M4
st,stm32mp1-m4-tee: load of a signed image on stm32mp1 Cortex-M4
st,stm32mp2-m33: Not supported ( need MMU configuration based on DT
                 reserved memories)
st,stm32mp2-m33-tee: load of a signed image on stm32mp2 Cortex-M33

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/u-boot.git 033ac89db8c1c063ef4b5b596b5c00ddd7cb200d remoteproc: add support of OP-TEE loading based on compatibility]
Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Change-Id: Id1735e492cb203640a18b308c9a9ccb573469994
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/u-boot/+/350378
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Patrice CHOTARD <patrice.chotard@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/u-boot/+/359777
Tested-by: Patrice CHOTARD <patrice.chotard@foss.st.com>
Domain-Review: Patrice CHOTARD <patrice.chotard@foss.st.com>
---
 drivers/remoteproc/stm32_copro.c | 36 +++++++++++++++++++-------------
 1 file changed, 21 insertions(+), 15 deletions(-)

diff --git a/drivers/remoteproc/stm32_copro.c b/drivers/remoteproc/stm32_copro.c
index 45eb3fd6b8e..bb6b3ec737a 100644
--- a/drivers/remoteproc/stm32_copro.c
+++ b/drivers/remoteproc/stm32_copro.c
@@ -68,22 +68,27 @@ static int stm32_copro_probe(struct udevice *dev)
 			return ret;
 	}
 
-	ret = rproc_optee_open(trproc);
-	if (!ret) {
-		dev_info(dev, "delegate the firmware management to OPTEE\n");
-		return 0;
-	}
+	if (device_is_compatible(dev, "st,stm32mp1-m4-tee") ||
+	    device_is_compatible(dev, "st,stm32mp2-m33-tee")) {
+		ret = rproc_optee_open(trproc);
+		if (ret) {
+			dev_err(dev, "failed to delegate to OP-TEE\n");
+			return ret;
+		}
 
-	ret = reset_get_by_name(dev, "mcu_rst", &priv->reset_ctl);
-	if (ret) {
-		dev_err(dev, "failed to get reset (%d)\n", ret);
-		return ret;
-	}
+		dev_info(dev, "delegate the firmware management to OP-TEE\n");
+	} else {
+		ret = reset_get_by_name(dev, "mcu_rst", &priv->reset_ctl);
+		if (ret) {
+			dev_err(dev, "failed to get reset (%d)\n", ret);
+			return ret;
+		}
 
-	ret = reset_get_by_name(dev, "hold_boot", &priv->hold_boot);
-	if (ret) {
-		dev_err(dev, "failed to get hold boot (%d)\n", ret);
-		return ret;
+		ret = reset_get_by_name(dev, "hold_boot", &priv->hold_boot);
+		if (ret) {
+			dev_err(dev, "failed to get hold boot (%d)\n", ret);
+			return ret;
+		}
 	}
 
 	dev_dbg(dev, "probed\n");
@@ -342,7 +347,8 @@ static const struct dm_rproc_ops stm32_copro_ops = {
 
 static const struct udevice_id stm32_copro_ids[] = {
 	{ .compatible = "st,stm32mp1-m4", .data = STM32MP15_M4_FW_ID },
-	{ .compatible = "st,stm32mp2-m33", .data = STM32MP25_M33_FW_ID },
+	{ .compatible = "st,stm32mp1-m4-tee", .data = STM32MP15_M4_FW_ID },
+	{ .compatible = "st,stm32mp2-m33-tee", .data = STM32MP25_M33_FW_ID },
 	{}
 };
 
-- 
2.47.2

